### PR TITLE
X-Forwarded-For

### PR DESCRIPTION
**Pull request by jefferai**
Docs are still todo.

This uses the approach of "specify how many hops back we should be looking" rather than some metric like "is a public IP" in order to validate. Behavior on unauthorized address or missing header can be set to reject or allow (without actually trusting the header).

### CODE REVIEW

File: command/server/config.go [hunk 1]
```diff
@@ -761,6 +761,10 @@ func parseListeners(result *Config, list *ast.ObjectList) error {
 			"address",
 			"cluster_address",
 			"endpoint",
+			"forwarded_for_authorized_addrs",
+			"forwarded_for_hop_skips",
+			"forwarded_for_reject_non_authorized",
```
Comment (vishalnayak) [inline command/server/config.go]: Thoughts on `non_authorized` vs `not_authorized`? `not_authorized` can go a little better with `not_present`.

File: helper/parseutil/parseutil.go [hunk 2]
```diff
@@ -118,3 +121,43 @@ func ParseCommaStringSlice(in interface{}) ([]string, error) {
...
+		}
+
+	default:
+		return nil, fmt.Errorf("unknown address input type %T", addrs)
+	}
+
+	for _, addr := range stringAddrs {
+		sa, err := sockaddr.NewSockAddr(addr)
+		if err != nil {
+			return nil, errwrap.Wrapf("error parsing address: {{err}}", err)
```
Comment (vishalnayak) [inline helper/parseutil/parseutil.go]: Do we want to mention the problematic `addr` in the error statement?

File: command/server/config.go (hunk 1 above)
Comment (jefferai) [reply command/server/config.go]: Good idea, changed.

File: helper/parseutil/parseutil.go (hunk 2 above)
Comment (jefferai) [reply helper/parseutil/parseutil.go]: Done.

File: http/handler.go [hunk 3]
```diff
@@ -124,6 +127,93 @@ func wrapGenericHandler(h http.Handler) http.Handler {
...
+		}
+
+		// At this point we have at least one value and it's authorized
+
+		// Split comma separated ones, which are common. This brings it in line
+		// to the multiple-header case.
+		if len(headers) == 1 {
+			headers = strings.Split(headers[0], ",")
+		}
+		for i, v := range headers {
```
Comment (calvn) [inline http/handler.go]: Is passing multiple `X-Forward-For` headers that are each comma separated valid [0]? If so, this is only iterating through the first set and not parsing the values correctly.

[0]: `-H 'X-Forwarded-For: 203.0.113.100, 203.0.113.101' -H 'X-Forwarded-For: 203.0.113.200, 203.0.113.201'`

Comment (jefferai) [reply http/handler.go]: It's not an actual standard, so...no idea? I guess I can handle that.

Comment (jefferai) [reply http/handler.go]: Fixed.

**Pull request merged**